### ISSUE TITLE
TypeError: Cannot read property 'symbol' of undefined

### ISSUE THREAD

**Issue by scriby** [sebastian-lenz/typedoc#70]
Do you have any ideas on the cause of this error? When I run the typescript compiler directly (the one included with typedoc), I don't receive an error. I'm using typedoc 0.2.3.

```
Using TypeScript 1.4.1 from .../node_modules/grunt-typedoc/node_modules/typedoc/node_modules/typescript/bin
.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1353
                            var declaration = new td.DeclarationReflection();
                                              ^
TypeError: Cannot read property 'symbol' of undefined
    at registerReflection (.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1128:42)
    at extractObjectType (.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1357:29)
    at extractType (.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1283:28)
    at .../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1214:50
    at Array.forEach (native)
    at .../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1212:73
    at withTypeParameters (.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1450:17)
    at createSignature (.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1205:17)
    at visitFunctionDeclaration (.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1842:41)
    at visit (.../node_modules/grunt-typedoc/node_modules/typedoc/bin/typedoc.js:1509:32)
```

**Comment by sebastian-lenz** [comment]
This looks like a bug in TypeDoc. At first glance it looks like an error while reflecting a type literal. Could you share the source file with me?

**Comment by SergioMorchon** [comment]
Same here, but with lots of propietary sourcecode without identify which of them are triggering the error :S
```
C:\Program Files\nodejs\node_modules\typedoc\bin\typedoc.js:1128
                var id = getSymbolID(node.symbol);
                                         ^
TypeError: Cannot read property 'symbol' of null
    at registerReflection (C:\Program Files\nodejs\node_modules\typedoc\bin\type
doc.js:1128:42)
    at extractObjectType (C:\Program Files\nodejs\node_modules\typedoc\bin\typed
oc.js:1357:29)
    at extractType (C:\Program Files\nodejs\node_modules\typedoc\bin\typedoc.js:
1283:28)
    at C:\Program Files\nodejs\node_modules\typedoc\bin\typedoc.js:1318:36
    at Array.forEach (native)
    at extractUnionType (C:\Program Files\nodejs\node_modules\typedoc\bin\typedo
c.js:1317:32)
    at extractType (C:\Program Files\nodejs\node_modules\typedoc\bin\typedoc.js:
1274:28)
    at extractObjectType (C:\Program Files\nodejs\node_modules\typedoc\bin\typed
oc.js:1341:34)
    at extractType (C:\Program Files\nodejs\node_modules\typedoc\bin\typedoc.js:
1283:28)
    at C:\Program Files\nodejs\node_modules\typedoc\bin\typedoc.js:1313:36
```

**Comment by scriby** [comment]
I looked into this a bit more today and was able to work around the errors by tweaking the source. (I added console.logs to visitSourceFile and extractObjectType to figure out which code was causing the problems).

All the problems dealt with trying to determine return types when they weren't explicitly declared.

The first problem was caused by something like this:

```
export interface A {
  x: () => {};
}
```

export function func(a: A) {
  return a.x();
}
```

I fixed that one by changing `x: () => {}` to `x: () => any`, which is what it should have been anyway.

I couldn't tell what was going on in the other instances, but I worked around them by explicitly setting the return type on the methods. Here is a code excerpt from one of the instances (getBreakpoints was the source of failure):

```
import $ = require('jquery');
'use strict';

export interface IBreakpointRange { start: number; end: number }

var _breakpoints: {
  small: IBreakpointRange;
  medium: IBreakpointRange;
  large: IBreakpointRange;
  xlarge: IBreakpointRange;
  xxlarge: IBreakpointRange;
};

export function getBreakpoints() {
  if (!_breakpoints) {
    var small = $('.foundation-mq-small').width();
    var medium = $('.foundation-mq-medium').width();
    var large = $('.foundation-mq-large').width();
    var xlarge = $('.foundation-mq-xlarge').width();
    var xxlarge = $('.foundation-mq-xxlarge').width();

    _breakpoints = {
      small: { start: small, end: medium },
      medium: { start: medium, end: large },
      large: { start: large, end: xlarge },
      xlarge: { start: xlarge, end: xxlarge },
      xxlarge: { start: xxlarge, end: 999999999 }
    };
  }

  return _breakpoints;
}
```

Thanks,

Chris

**Comment by sebastian-lenz** [comment]
Thank's for taking your time to investigate in this error. I've added your code producing the error as a test case. It seems to be resolved in the latest version of TypeDoc.

**Comment by danielearwicker** [comment]
I'm itching to use typedoc but I hit this:

    c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1128
                    var id = getSymbolID(node.symbol);
                                             ^
    TypeError: Cannot read property 'symbol' of undefined
        at registerReflection (c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1128:42)
        at extractObjectType (c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1357:29)
        at extractType (c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1283:28)
        at c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1214:50
        at Array.forEach (native)
        at c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1212:73
        at withTypeParameters (c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1450:17)
        at createSignature (c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1205:17)
        at visitFunctionDeclaration (c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1842:41)
        at visit (c:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1509:32)

Looks very similar to this issue. Being keen to try the latest code, I tried cloning the repository just now, followed by the usual npm install, grunt, and then ran bin/typedoc instead, but now I get:

    c:\Source\Git\github\typedoc\bin\typedoc.js:6383
                            this.theme = new themeClass(this, path);
                                         ^
    TypeError: undefined is not a function
        at Renderer.prepareTheme (c:\Source\Git\github\typedoc\bin\typedoc.js:6383:38)
        at Renderer.getParameters (c:\Source\Git\github\typedoc\bin\typedoc.js:6268:22)
        at Application.collectParameters (c:\Source\Git\github\typedoc\bin\typedoc.js:255:47)
        at Application.bootstrapFromCommandline (c:\Source\Git\github\typedoc\bin\typedoc.js:103:18)
        at new Application (c:\Source\Git\github\typedoc\bin\typedoc.js:79:22)
        at Object.<anonymous> (c:\Source\Git\github\typedoc\bin\typedoc:4:12)
        at Module._compile (module.js:456:26)
        at Object.Module._extensions..js (module.js:474:10)
        at Module.load (module.js:356:32)
        at Function.Module._load (module.js:312:12)

Specifying --theme default didn't change it. Is there anything else I need to do to get the latest code working?

**Comment by sebastian-lenz** [comment]
Hi Daniel, in order to use the current version you also need to update the dependency <code>typedoc-default-themes</code>.

However, I've pushed v0.3 to npm right now, so give it a try :wink:

**Comment by danielearwicker** [comment]
I got a whole new error message with 0.3:

    undefined:584
                    throw new Error("Debug Failure. False expression: " + (message
                          ^
    Error: Debug Failure. False expression:
        at Object.assert (eval at <anonymous> (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:41:3)
    , <anonymous>:584:23)
        at isContextSensitive (eval at <anonymous> (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:
    41:3), <anonymous>:12976:22)
        at resolveLocation (eval at <anonymous> (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:41:
    3), <anonymous>:14000:84)
        at Object.getTypeAtLocation (eval at <anonymous> (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typed
    oc.js:41:3), <anonymous>:14013:13)
        at extractSignatureType (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:3277:36)
        at C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:3258:34
        at Context.withScope (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1767:17)
        at Object.createSignature (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:3254:21)
        at C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:2533:47
        at Context.withScope (C:\Users\Daniel\AppData\Roaming\npm\node_modules\typedoc\bin\typedoc.js:1767:17)

**Comment by sebastian-lenz** [comment]
Okay, that's ugly, this error is raised by TypeScript. The stack trace is not very meaningful to me, it's reflecting a method/function/function definition or something like that and I cannot reproduce this issue with my test cases.

I've included a better error logging in the current master branch. It would be of great help if you could fetch the current version and run it against you project. The error message should include the related source code and this time the themes should be up to date.

**Comment by danielearwicker** [comment]
Sure, here is the output:

    Using TypeScript 1.4.1 from c:\source\git\github\typedoc\node_modules\typescript\bin
    Error: An error occurred while creating reflections for the current project.
     | Please report this error at https://github.com/sebastian-lenz/typedoc/issues
     | including the following details:
     |
     | The error occurred while converting `__object.remove` in `colorPalette.ts` around line 73:
     |  >             code: ko.observable(code || 'rgba(0,0,0,1)'),
     |  @             name: nameObs,
     |  >             remove() {
     |  >                 this.colors.remove(color);
     |  >             },
     |  >             nameValidationError: ko.pureComputed(() =>
     |  >                 (!nameObs()) ? localisation.get('errors_colorPalette_nameMissing')
     |  >                 : (this.colors().filter(color => color.name() === nameObs()).length == 1) ? ''
     |  >                 : localisation.get('errors_colorPalette_duplicateName'))
     |  >         }
     |  >         return color;
     |  >     }
     |  >
     |  >     private defaultColorName() {
     |  >         var colorNameString = localisation.get('color');
     |  >         var defName = colorNameString + 1,
     |  >             i = 1,
     |  >             colorNames = this.colors().map(color => color.name());
     |  >         while (colorNames.some(e => e == defName)) {
     |  >             defName = colorNameString + i++;
     |  >         }
     |  >         return defName;
     |  >     }
     |  >
     |  >     load(data: any) {
     |  >         this.colors.removeAll();
     |  >         Object.keys(data.colors).sort().forEach(name => {
     |  >             var code = data.colors[name];
     |
     | Error: Debug Failure. False expression:
     |   at Object.assert (eval at <anonymous> (c:\source\git\github\typedoc\bin\typedoc.js:41:3), <anonymous>:584:23)
     |   at isContextSensitive (eval at <anonymous> (c:\source\git\github\typedoc\bin\typedoc.js:41:3), <anonymous>:12976:22
    )
     |   at resolveLocation (eval at <anonymous> (c:\source\git\github\typedoc\bin\typedoc.js:41:3), <anonymous>:14000:84)
     |   at Object.getTypeAtLocation (eval at <anonymous> (c:\source\git\github\typedoc\bin\typedoc.js:41:3), <anonymous>:14
    013:13)
     |   at extractSignatureType (c:\source\git\github\typedoc\bin\typedoc.js:3317:36)
     |   at c:\source\git\github\typedoc\bin\typedoc.js:3298:34
     |   at Context.withScope (c:\source\git\github\typedoc\bin\typedoc.js:1775:17)
     |   at Object.createSignature (c:\source\git\github\typedoc\bin\typedoc.js:3294:21)
     |   at c:\source\git\github\typedoc\bin\typedoc.js:2573:47
     |   at Context.withScope (c:\source\git\github\typedoc\bin\typedoc.js:1775:17)

    Error: An error occurred while creating reflections for the current project.
     | Please report this error at https://github.com/sebastian-lenz/typedoc/issues
     | including the following details:
     |
     | The error occurred while converting `__object.nameValidationError` in `colorPalette.ts` around line 76:
     |  >                 this.colors.remove(color);
     |  @             },
     |  >             nameValidationError: ko.pureComputed(() =>
     |  >                 (!nameObs()) ? localisation.get('errors_colorPalette_nameMissing')
     |  >                 : (this.colors().filter(color => color.name() === nameObs()).length == 1) ? ''
     |  >                 : localisation.get('errors_colorPalette_duplicateName'))
     |  >         }
     |  >         return color;
     |  >     }
     |  >
     |  >     private defaultColorName() {
     |  >         var colorNameString = localisation.get('color');
     |  >         var defName = colorNameString + 1,
     |  >             i = 1,
     |  >             colorNames = this.colors().map(color => color.name());
     |  >         while (colorNames.some(e => e == defName)) {
     |  >             defName = colorNameString + i++;
     |  >         }
     |  >         return defName;
     |  >     }
     |  >
     |  >     load(data: any) {
     |  >         this.colors.removeAll();
     |  >         Object.keys(data.colors).sort().forEach(name => {
     |  >             var code = data.colors[name];
     |  >             this.colors.push(this.colorDic(name, code));
     |  >         });
     |  >         this.colors().some(c => {
     |
     | Error: Expected container reflection.
     |   at Object.createDeclaration (c:\source\git\github\typedoc\bin\typedoc.js:3160:23)
     |   at visitVariableDeclaration (c:\source\git\github\typedoc\bin\typedoc.js:2483:38)
     |   at visit (c:\source\git\github\typedoc\bin\typedoc.js:2224:32)
     |   at c:\source\git\github\typedoc\bin\typedoc.js:2647:21
     |   at Array.forEach (native)
     |   at visitObjectLiteral (c:\source\git\github\typedoc\bin\typedoc.js:2646:33)
     |   at Object.visit (c:\source\git\github\typedoc\bin\typedoc.js:2248:32)
     |   at c:\source\git\github\typedoc\bin\typedoc.js:2833:31
     |   at Array.forEach (native)
     |   at c:\source\git\github\typedoc\bin\typedoc.js:2832:37

    Error: c:/Source/TFS/trunk/Core/WebService/WebService/src/platform/base64.ts(63)
     Cannot compile external modules unless the '--module' flag is provided.

**Comment by sebastian-lenz** [comment]
Thanks for sharing this. Did you remove the last error? You must specify the <code>--module</code> parameter in order to allow TypeScript to compile your project. Does the problem persists afterwards?

I've created a simple test case that mimics your scenario and it works for me:
```
export function createSomething() {
    var result = {
        foo: 'bar',
        doSomething: (a:number) => a + 1,
        doAnotherThing: () => { }
    };

    return result;
}
```

**Comment by danielearwicker** [comment]
The last error goes away when I specify `--module commonjs`. However the other errors remain exactly the same. I am specifying an input folder that is the ultimate container of all my .ts files - is there anything else I need to do?

**Comment by sebastian-lenz** [comment]
No, there are no other mandatory arguments. Is your project available on GitHub so I could check it out? If not could maybe you could share the source of the function that is emitting the error?

**Comment by sebastian-lenz** [comment]
Is your code correct or did you miss the colon after ``remove``? Shouldn't the remove function be like that:
<pre><code>  remove: () => {
    this.colors.remove(color);
  }
</code></pre>

**Comment by danielearwicker** [comment]
No, that's valid TS, it's an ES6 style abbreviation (try it in the TS playground). This is a codebase of 56000 lines that builds without errors.

**Comment by danielearwicker** [comment]
> Is your project available on GitHub so I could check it out? 

Sadly not, it's proprietary code belonging to my employers. I will try to make a minimal repro of the error.

**Comment by sebastian-lenz** [comment]
Cool, didn't know about the ES6 notation.

I've found the cause of the error. The error is triggered when my example function above is placed inside a class and when using your type of function notation:

```
class Foo {
    createSomething() {
        return {
            foo: 'bar',
            doSomething(a:number) { return a + 1},
            doAnotherThing() {}
        };
    }
}
```

**Comment by sebastian-lenz** [comment]
I've pushed v0.3.1 to npm, hope it resolves this issue.

**Comment by danielearwicker** [comment]
Yes, it generated docs without errors! :+1: 

The index.html page is basically empty on the left, and on the right there's a Globals column containing a flat list of all my file paths. Is that right?

I'd like a hierarchical map of some kind, e.g. global-internal modules overview, external module folder structure, so it's more immediately browsable. Let me know if there isn't a way to do that already I'll try and see if I can hack it together to show what I mean.

**Comment by sebastian-lenz** [comment]
Great to hear. TypeDoc places your <code>readme.md</code> on the index page, so I guess you have an empty one flying around?

If you click on "Globals" in the navigation bar you should see a grouped index of your project, something like [this](http://typedoc.io/themes/default/globals.html). You can make this your default index page by specifying the parameter <code>--readme none</code>.

**Comment by sebastian-lenz** [comment]
@scriby: Is the original issue resolved with v0.3.1?

**Comment by sebastian-lenz** [comment]
I think this issue should be resolved with v0.3.1. Please message me if I'm wrong.

**Comment by vtkalek** [comment]
Hi,
I also found a bug for you, check this out:
I found that TypeDoc is falling with error message:

"  interfaceReflection.children.forEach(function.....
TypeError: Cannot read property 'forEach' of undefined.... "

It happens when it runs on the definition of such type:

" ....
   export interface Entity {
        //No members
    } 
.....
"

So there should be a check if interfaceReflection really has children, something like if(interfaceReflection.children)

Best regards, Alex
